### PR TITLE
Add wrapper for agglomerateByRank/mergeRows

### PR DESCRIPTION
**Pull request by 0xMuluh**
There are 2 different methods doing similar (merging/grouping rows/features);
- mergeRows and agglomerateByRank
This wrapper method that combines these two methods. 
- It would be used internally by other functions so it would not be available directly for user.

### CODE REVIEW

File: R/utils.R [hunk 1]
```diff
@@ -283,3 +283,19 @@
     
     taxa_tab
 }
+
+################################################################################
+# internal wrappers for agglomerateByRank/mergeRows
+.merge_features <- function(x, merge.by, ..., agglomerateTree = FALSE) {
```
Comment (TuomasBorman) [inline R/utils.R]: is `agglomerateTree` necessary here? Can it be passed similarly to other arguments with `...`?

**Review by TuomasBorman — APPROVED** [review]
Neat!

Add couple of checks: .merge_features vs agglomerateByRank and .merge_features vs mergeRows

File: R/utils.R [hunk 2]
```diff
@@ -286,16 +286,18 @@
...
+    if (is.character(merge.by) && length(merge.by) < 2 && merge.by %in% taxonomyRanks(x)) {
+         #Merge using agglomerateByRank
+        x <- agglomerateByRank(x, rank = merge.by, ...)
     } else {
         # Merge using mergeRows
         x <- mergeRows(x, f = merge.by, ...)
     }
     
     return(x)
+
```
Comment (TuomasBorman) [inline R/utils.R]: You can remove this additional lines

File: R/utils.R [hunk 3]
```diff
@@ -286,16 +286,18 @@
 
 ################################################################################
 # internal wrappers for agglomerateByRank/mergeRows
-.merge_features <- function(x, merge.by, ..., agglomerateTree = FALSE) {
+.merge_features <- function(x, merge.by, ...) {
     # Check if merge.by parameter belongs to taxonomyRanks
-    if (merge.by %in% taxonomyRanks(x)) {
-        # Merge using agglomerateByRank
-        x <- agglomerateByRank(x, rank = merge.by, ..., agglomerateTree = agglomerateTree)
+    if (is.character(merge.by) && length(merge.by) < 2 && merge.by %in% taxonomyRanks(x)) {
```
Comment (TuomasBorman) [inline R/utils.R]: length(merge.by) == 1 would be more explicit

File: R/utils.R [hunk 4]
```diff
@@ -286,16 +286,18 @@
...
     } else {
         # Merge using mergeRows
         x <- mergeRows(x, f = merge.by, ...)
     }
     
     return(x)
+
 }
 
+
```
Comment (TuomasBorman) [inline R/utils.R]: this also

File: tests/testthat/test-2merge.R [hunk 5]
```diff
@@ -102,11 +102,14 @@ test_that("merge", {
     merged  <- mergeRows(tse, f=rowData(tse)$group2, mergeTree=TRUE)
     merged2 <- mergeRows(tse, f = rowData(tse)$group2, mergeTree = FALSE)
     merged3 <- mergeRows(esophagus, f = rowData(esophagus)$group2, mergeTree = TRUE)
+    merged4 <- .merge_features(tse, merge.by = rowData(tse)$group2)
+    merged5 <- mergeRows(tse, f = rowData(tse)$group2)
```
Comment (TuomasBorman) [inline tests/testthat/test-2merge.R]: How about trying to pass e.g. `mergeTree` parameter? You could add `mergeTree = TRUE` here to ensure that `...` is working

File: tests/testthat/test-3agglomerate.R [hunk 6]
```diff
@@ -51,6 +51,11 @@ test_that("agglomerate", {
     expect_equal(rowData(actual)$Family,c("c",NA,"d","e","f","g","h",NA))
     actual <- agglomerateByRank(xtse,"Phylum")
     expect_equivalent(rowData(actual),rowData(actual_phylum))
+    
+    #
```
Comment (TuomasBorman) [inline tests/testthat/test-3agglomerate.R:54]: You could remove extra line

**Review by TuomasBorman — CHANGES_REQUESTED** [review]
Couple small thhings, looks good!

File: R/utils.R (hunk 1 above)
Comment (0xMuluh) [reply R/utils.R]: ack

File: R/utils.R (hunk 2 above)
Comment (0xMuluh) [reply R/utils.R]: ack

File: R/utils.R (hunk 4 above)
Comment (0xMuluh) [reply R/utils.R]: ack

File: R/utils.R (hunk 3 above)
Comment (0xMuluh) [reply R/utils.R]: thanks:)

File: tests/testthat/test-2merge.R (hunk 5 above)
Comment (0xMuluh) [reply tests/testthat/test-2merge.R]: ack

File: tests/testthat/test-3agglomerate.R (hunk 6 above)
Comment (0xMuluh) [reply tests/testthat/test-3agglomerate.R:54]: ack

**Review by TuomasBorman — APPROVED** [review]
very nice

**Comment by antagomir** [comment]
Status of this?

**Comment by 0xMuluh** [comment]
> Status of this?

Complete for now.

**Comment by antagomir** [comment]
Great!

We had elsewhere discussion about function naming. 

I would suggest to update as follows:

- mergeRows -> mergeFeatures (see issue #392)
- agglomerateByRank -> mergeFeaturesByRank (logical to use similar naming in both functions)

Downside of "merge" term is that it is also used in another meaning, to combine full SE objects, data.frames, matrices etc (instead of within object features as here). Another option would be: agglomerateFeatures and agglomerateFeaturesByRank but that is slower to write..  or combineFeature / combineFeaturesByRank (but combine is also used in the same way than merge, between distinct DataFrames)

Shall we deal with these issues here in this closely related PR, or should we open a new separate issue & PR?

**Comment by TuomasBorman** [comment]
> Great!
> 
> We had elsewhere discussion about function naming.
> 
> I would suggest to update as follows:
> 
>     * mergeRows -> mergeFeatures (see issue [Deprecate mergeRows/mergeCols #392](https://github.com/microbiome/mia/issues/392))
> 
>     * agglomerateByRank -> mergeFeaturesByRank (logical to use similar naming in both functions)
> 
> 
> Downside of "merge" term is that it is also used in another meaning, to combine full SE objects, data.frames, matrices etc (instead of within object features as here). Another option would be: agglomerateFeatures and agglomerateFeaturesByRank but that is slower to write.. or combineFeature / combineFeaturesByRank (but combine is also used in the same way than merge, between distinct DataFrames)
> 
> Shall we deal with these issues here in this closely related PR, or should we open a new separate issue & PR?

I also think `mergeFeatures` and `mergeFeaturesByRank` are better than `agglomerate*`, and it is good to have same naming convention.

We discussed also about enabling `rowData` variables in `mergeRows`. Currently, the groups must be fed as a vector (`f` parameter) but I think it would be nice to be able to specify the grouping variable directly from `rowData` as a column name. (Same for mergeCols)

There must be some reason why we didn't implement this before??? Do you remember @antagomir?

I think this PR can be merged, and these new modifications can be done in different PR. @Daenarys8 can you implement these?

**Comment by antagomir** [comment]
I agree. I don't think there is a specific reason, the rowData variables could be called by name as well.

If @Daenarys8 can open a new issue and PR about that it would be great.

You can close this issue when you have checked that it is ready.

**Comment by 0xMuluh** [comment]
> > Great!
> > We had elsewhere discussion about function naming.
> > I would suggest to update as follows:
> > ```
> > * mergeRows -> mergeFeatures (see issue [Deprecate mergeRows/mergeCols #392](https://github.com/microbiome/mia/issues/392))
> > 
> > * agglomerateByRank -> mergeFeaturesByRank (logical to use similar naming in both functions)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Downside of "merge" term is that it is also used in another meaning, to combine full SE objects, data.frames, matrices etc (instead of within object features as here). Another option would be: agglomerateFeatures and agglomerateFeaturesByRank but that is slower to write.. or combineFeature / combineFeaturesByRank (but combine is also used in the same way than merge, between distinct DataFrames)
> > Shall we deal with these issues here in this closely related PR, or should we open a new separate issue & PR?
> 
> I also think `mergeFeatures` and `mergeFeaturesByRank` are better than `agglomerate*`, and it is good to have same naming convention.
> 
> We discussed also about enabling `rowData` variables in `mergeRows`. Currently, the groups must be fed as a vector (`f` parameter) but I think it would be nice to be able to specify the grouping variable directly from `rowData` as a column name. (Same for mergeCols)
> 
> There must be some reason why we didn't implement this before??? Do you remember @antagomir?
> 
> I think this PR can be merged, and these new modifications can be done in different PR. @Daenarys8 can you implement these?

Sure,  will give it a shot

**Comment by antagomir** [comment]
Can someone close this PR when it is confirmed to be ready.

**Comment by antagomir** [comment]
Was this ever merged? If I check correctly from above, it was just closed without merging ? The idea was to merge I guess?

**Comment by antagomir** [comment]
Hmm - - ok now I noticed that the renaming scheme discussed above, and agreed, has not made it to this PR yet.

Can we add it (@Daenarys8) ? 

More specifically, the idea was to rename "Rows" to "Features" and "Cols" to "Samples".

In addition, to harmonize terminology (use "merge" instead of "agglomerate").

However, one last thing to discuss first: the meaning of "to agglomerate" is somewhat better fit with our case (in terms of language & meaning), in particular when the phylogenetic tree is involved in the process. The phyloseq equivalent to TreeSE `agglomerateByRank` is `tax_glom` (e.g. https://mikemc.github.io/speedyseq/reference/tax_glom.html).

We could use "glom" instead of "merge". That would be as fast to write, and the meaning would be more specific (merge is easier to confuse with merging of matrices). However, "glom" might be a bit weird term to introduce right now, and it is possible to re-evaluate that later as well.

Hence, in summary, I suggest to rename as:
- mergeCols -> mergeSamples (or glomSamples... perhaps not?)
- mergeRows -> mergeFeatures
- agglomerateByRank -> mergeFeaturesByRank 
- agglomerateByPrevalence -> mergeFeaturesByPrevalence

@Daenarys8 could you add that to this PR as discussed above, OR open a new issue proposing this change, then merging and closing the current PR. Also confirm that the other points discussed / agreed above have now been addessred.

**Comment by antagomir** [comment]
Ok it is more clear to do the other discussed changes in a separate PR. I am merging and closing this one.

**Comment by antagomir** [comment]
Ok - now as this is merged:

the original motivation was ANCOMBC issue #174 where we would like to let users specify taxonomic rank or also other rowData variable to merge rows before running ANCOMBC: https://github.com/FrederickHuangLin/ANCOMBC/issues/174

Now, ideally this new wrapper will help there; it would group by taxonomic rank if this is available in rowData, otherwise it uses the more general grouping.

**Pull request merged**